qiskit/providers/ibmq/api/websocket.py

```diff
@@ -0,0 +1,148 @@
+# -*- coding: utf-8 -*-
+
+# Copyright 2018, IBM.
+#
+# This source code is licensed under the Apache License, Version 2.0 found in
+# the LICENSE.txt file in the root directory of this source tree.
+
+"""Utilities for IBM Q API connector."""
+
+import asyncio
+import json
+import time
+from concurrent import futures
+
+import websockets
+
+from .exceptions import WebsocketError, WebsocketTimeoutError
+
+
+class WebsocketMessage:
+    """Container for a message sent or received via websockets.
+
+    Attributes:
+        type_ (str): message type.
+        data (dict): message data.
+    """
+    def __init__(self, type_, data=None):
+        self.type_ = type_
+        self.data = data
+
+    def as_json(self):
+        """Return a json representation of the message."""
+        parsed_dict = {'type': self.type_}
+        if self.data:
+            parsed_dict['data'] = self.data
+        return json.dumps(parsed_dict)
+
+    @classmethod
+    def from_bytes(cls, json_string):
+        """Instantiate a message from a bytes response."""
+        parsed_dict = json.loads(json_string.decode('utf8'))
+        return cls(parsed_dict['type'], parsed_dict.get('data', None))
+
+
+class WebsocketClient:
+    """Client for websocket communication with the IBMQ API.
+
+    Attributes:
+        websocket_url (str): URL for websocket communication with IBM Q.
+        access_token (str): access token for IBMQ.
+    """
+
+    def __init__(self, websocket_url, access_token):
+        self.websocket_url = websocket_url
+        self.access_token = access_token
+
+    @asyncio.coroutine
+    def connect(self, url):
+        """Authenticate against the websocket server, returning the connection.
+
+        Returns:
+            Connect: an open websocket connection.
+
+        Raises:
+            WebsocketError: if the connection to the websocket server failed
+                due to connection or authentication errors.
+        """
+        websocket = yield from websockets.connect(url)
+
+        try:
+            # Authenticate against the server.
+            auth_request = self._authentication_message()
+            yield from websocket.send(auth_request.as_json())
+
+            # Verify that the server acknowledged our authentication.
+            auth_response_raw = yield from websocket.recv()
+            auth_response = WebsocketMessage.from_bytes(auth_response_raw)
+
+            if auth_response.type_ != 'authenticated':
+                raise WebsocketError(auth_response.as_json())
+        except websockets.ConnectionClosed as ex:
+            yield from websocket.close()
+            raise WebsocketError('Error in websocket authentication: %s' % ex)
+
+        return websocket
+
+    @asyncio.coroutine
+    def get_job_status(self, job_id, timeout=None):
```

Comment:
We'll revise in the next PR, as the main reason for the name was to keep it in line with the `connector` naming (ie. since this methods uses the endpoint `{}/jobs/{}/status` -> `get_job_status()`) and agnostic to its usage - hopefully it will become clear one way or the other when everything is fully in place.